### DISCUSSION TITLE
Looking for a way to convert constraints into a C/C++ for compilation on ARM

### DISCUSSION THREAD

**Discussion by andreyb051** [fvutils/pyvsc#74] [General]
Hi,

Has anybody looked into it? My test suite runs on ARM, without real operating system.
So I would like to compile constrained randomization code into C/C++, then compile it into my test suite, for overnight executions. Anybody tried something of the sort?
Again, no operating system, no multi-threading, no ability to run python, just C/C++.

Thanks,
   Andrey

**Comment by mballance** [comment]
This has been done a few ways, though there aren't any completely-general solutions that I'm aware of. 

It's not uncommon to use a constrained-random generation like SystemVerilog or PyVSC to pre-generate software (assembly, C/C++) that is then compiled and run on the processor core. One example, here, is the [RISCV-DV](https://github.com/google/riscv-dv) tool that generates RISC-V assembly-level test programs. You could use a similar approach to generate C code as well. Depending on how fast your execution platform is (simulation, FPGA, etc), you might need to pre-generate a suite of test programs to fill an overnight run.

[IBM Research](https://researcher.watson.ibm.com/researcher/view_group_pubs.php?grp=1347) has done some work at creating long-running test programs that do randomization within the generated test itself. The tradeoff appears to be that the targeted type of tests are very domain-specific, not general constraints.

**Comment by andreyb051** [comment]
Thanks!

**Comment by qzcx** [comment]
At my work, we generate C/C++ test parameters from using python on our host and then run the compiled C/C++ on the embedded processor. 

Useful python libraries are:
 - ctypes. Generating C headers from python. Write out C struct binaries from python to load directly in C. 
 - pycparser. parse existing c header files for use in python. I wrote some functions to generate python classes from existing headers with this.
 - pyvsc for constrained random :) Still in "proof of concept" stage for us
